KAFKA-18914 Migrate ConsumerRebootstrapTest to use new test infra

clarkwtc opened the pull request:
Migrate ConsumerRebootstrapTest to the new test infra and remove the old Scala test.

The PR changed three things.
* Migrated `ConsumerRebootstrapTest` to new test infra and removed the old Scala test.
* Updated the original test case to cover rebootstrap scenarios.
* Integrated `ConsumerRebootstrapTest` into `ClientRebootstrapTest` in the `client-integration-tests` module.
* Removed the `RebootstrapTest.scala`.

Default `ConsumerRebootstrap` config:
> properties.put(CommonClientConfigs.METADATA_RECOVERY_STRATEGY_CONFIG, "rebootstrap");
       properties.put(CommonClientConfigs.METADATA_RECOVERY_REBOOTSTRAP_TRIGGER_MS_CONFIG, "300000");
       properties.put(CommonClientConfigs.SOCKET_CONNECTION_SETUP_TIMEOUT_MS_CONFIG, "10000");
       properties.put(CommonClientConfigs.SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS_CONFIG, "30000");
       properties.put(CommonClientConfigs.RECONNECT_BACKOFF_MS_CONFIG, "50L");
       properties.put(CommonClientConfigs.RECONNECT_BACKOFF_MAX_MS_CONFIG, "1000L");

The test case for the consumer with enabled rebootstrap
![Screenshot 2025-03-22 at 9 48 13 PM](https://github.com/user-attachments/assets/8470549f-a24c-43fa-ae44-789cbf422a63)


The test case for the consumer with disabled rebootstrap
![Screenshot 2025-03-22 at 9 47 22 PM](https://github.com/user-attachments/assets/0a183464-6a74-449f-8e71-d641a6ea5bb1)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

Commit:
feat: Migrate ConsumerRebootstrapTest to use new test infra

Commit:
feat: Remove old `ConsumerRebootstrapTest` test

Commit:
Merge branch 'trunk' into KAFKA-18914-rewrite-consumer_rebootstrap_test

chia7712 commented:
@clarkwtc could you please merge it into ClientRebootstrapTest?

Commit:
Merge branch 'trunk' into KAFKA-18914-rewrite-consumer_rebootstrap_test

Commit:
feat: Move this test to clients-integration-tests

clarkwtc commented:
@chia7712 
I merged it into ClientRebootstrapTest.

Commit:
feat: Cover rebootstrap of classic consumer test cases.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -94,4 +101,129 @@ public void testAdminRebootstrapDisabled(ClusterInstance clusterInstance) {
+            try (var producer = clusterInstance.producer()) {
+                var recordMetadata = producer.send(new ProducerRecord<>(TOPIC, part, "key 1".getBytes(), "value 1".getBytes())).get();
+                assertEquals(1, recordMetadata.offset());
+                producer.flush();
```
we don't need to call `flush` after you call the `get`

clarkwtc replied:
Maybe we should remove to call `get`?
I need to call `flush`; without it, if the broker restarts on rebootstrap, the consumer polls nothing.

chia7712 replied:
Given that `acks` is not set to 0, the `get` method guarantees that the record is transmitted to the server. This behavior is equivalent to a flush operation.

clarkwtc replied:
Sure, set `acks=-1` to guarantee that the record will not be lost.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -94,4 +101,129 @@ public void testAdminRebootstrapDisabled(ClusterInstance clusterInstance) {
+        var tp = new TopicPartition(TOPIC, part);
+
+        try (var producer = clusterInstance.producer()) {
+            var recordMetadata = producer.send(new ProducerRecord<>(TOPIC, part, "key 0".getBytes(), "value 0".getBytes())).get();
```
there is only one partition, so we can streamline the code:
```
var recordMetadata = producer.send(new ProducerRecord<>(TOPIC, "value 0".getBytes())).get();
```

clarkwtc replied:
Got it; make it simple.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -94,4 +101,129 @@ public void testAdminRebootstrapDisabled(ClusterInstance clusterInstance) {
+            clusterInstance.startBroker(broker0);
+
+            try (var producer = clusterInstance.producer()) {
+                var recordMetadata = producer.send(new ProducerRecord<>(TOPIC, part, "key 1".getBytes(), "value 1".getBytes())).get();
```
ditto

clarkwtc replied:
I fixed it.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -94,4 +101,129 @@ public void testAdminRebootstrapDisabled(ClusterInstance clusterInstance) {
+            // Only the server 1 is available for the consumer during the bootstrap.
+            consumer.assign(partitions);
+            consumer.seekToBeginning(partitions);
+            assertEquals(1, consumer.poll(Duration.ofSeconds(1)).count());
```
we can't assume the records get returned by only one poll. could you please add loop to avoid flaky?

clarkwtc replied:
Okay, I use `TestUtils.waitForCondition` to prevent flaky.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -94,4 +101,129 @@ public void testAdminRebootstrapDisabled(ClusterInstance clusterInstance) {
+
+            // The server 1 originally cached during the bootstrap, is offline.
+            // However, the server 0 from the bootstrap list is online.
+            assertEquals(1, consumer.poll(Duration.ofSeconds(1)).count());
```
ditto

clarkwtc replied:
I fixed it.

chia7712 commented on clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/ClientRebootstrapTest.java:
```diff
@@ -94,4 +101,129 @@ public void testAdminRebootstrapDisabled(ClusterInstance clusterInstance) {
+        var partitions = List.of(new TopicPartition(TOPIC, part));
+
+        try (var producer = clusterInstance.producer()) {
+            var recordMetadata = producer.send(new ProducerRecord<>(TOPIC, part, "key 0".getBytes(), "value 0".getBytes())).get();
```
ditto

clarkwtc replied:
I fixed it.

chia7712 reviewed (commented):
@clarkwtc thanks for this patch.

Commit:
fix: Make it be better

- Add ack equal 1 writing the record to local log.
- Wait consumer to poll record, that avoid flaky.
- Streamline ProducerRecord because only one partition.

Commit:
Merge branch 'trunk' into KAFKA-18914-rewrite-consumer_rebootstrap_test

Commit:
fix: Set `acks=all` for the leader wait setting of in-sync replicas to acknowledge the record.

Commit:
chore: Streamline variable part in TopicPartition.

Commit:
Merge branch 'trunk'

Commit:
feat: Remove unused `RebootstrapTest.scala`

Commit:
chore: Set `PARTITIONS` to global variable in `ClientRebootstrapTest`
